README.md
# lks-indicator

lks-indicator is an indicator app for Linux that displays status of Lock keys (Scroll, Caps and Num) in the panel.

It was originally written for Ubuntu Unity, and has been tested on Ubuntu 16.04 and Ubuntu Kylin 14.04, and Xubuntu 16.04.

Still, it will run fine on any desktop/distro that supports AppIndicator.

## Motivation for the project

Originally this had been written as an answer for ByteCommander's question at [AskUbuntu](http://askubuntu.com): [Keyboard LED panel indicator for Unity Desktop](http://askubuntu.com/q/796985/295286). Eventually it has been decided to turn it into standalone project.

## Demo
GIF demonstration of lks-indicator:

[![GIF demonstration of lks-indicator](http://i.imgur.com/FxSae8c.gif)](http://i.imgur.com/FxSae8c.gif)

## Installation

### [![Ubuntu](https://www.pylint.org/assets/img/ubuntu.png)](https://ubuntu.com) Ubuntu and Debian

Debian/Ubuntu `deb` package: [Download `lks-indicator.deb`](https://github.com/SergKolo/lks-indicator/raw/master/debian/lks-indicator.deb)

### Manual

Run the included install script:

    sudo sh install.sh

And to uninstall:

    sudo sh install.sh --uninstall

## Usage

When none of Scroll, Caps or Num lock are pressed, the indicator will be green.

When one or more lock keys are active, the indicator becomes red, and the following is displayed:

- *\[C\]* for *Caps Lock*
- *\[N\]* for *Num Lock*
- *\[S\]* for *Scroll Lock*

The indicator menu has only one option - closing the indicator (*Quit*).

The program has an option, `--show-all`, which makes it run so:

[![show-all no key pressed](http://i.imgur.com/LFmhJQ8.png)](http://i.imgur.com/LFmhJQ8.png)


[![show-all numlock on](http://i.imgur.com/CUnnsbh.png)](http://i.imgur.com/CUnnsbh.png)

## License

The indicator is provided under MIT license, copy of which is included with the source code


debian/changelog
lks-indicator (1.0) RELEASED; urgency=medium

  * Initial release. Builds from git.

 -- Sergiy Kolodyazhnyy <1047481448@qq.com> Sun, 17 Jul 2016 13:36:43 +0530


__init__.py
#!/usr/bin/env python
#-*- coding: utf-8-*-
#
# The MIT License (MIT)
# 
# Copyright (c) 2016 Sergiy Kolodyazhnyy <1047481448@qq.com>
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# 



import lock_keys_status


if __name__ == '__main__':
    try:
      lock_keys_status.main()
    except KeyboardInterrupt:
      pass



lock_keys_status.py
#!/usr/bin/env python
# coding: utf-8

###########################################################
# Author: Serg Kolo , contact: 1047481448@qq.com
# Date: July 16, 2012
# Purpose: Simple indicator of Caps, Num, and Scroll Lock
#		   keys for Ubuntu
#
# Written for: http://askubuntu.com/q/796985/295286
# Tested on: Ubuntu 16.04 LTS
#
###########################################################
#
# Licensed under The MIT License (MIT).
# See included LICENSE file or the notice below.
#
# Copyright © 2016 Sergiy Kolodyazhnyy
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import glib
import subprocess
import appindicator
import gtk
import os

class LockKeyStatusIndicator(object):

	def __init__(self):

		self.app = appindicator.Indicator('LKS', '',
						  appindicator.CATEGORY_APPLICATION_STATUS)

		self.app.set_status(appindicator.STATUS_ACTIVE)

		self.update_label()

		self.app_menu = gtk.Menu()
		self.quit_app = gtk.MenuItem('Quit')
		self.quit_app.connect('activate', self.quit)
		self.quit_app.show()
		self.app_menu.append(self.quit_app)

		self.app.set_menu(self.app_menu)


	def run(self):

		try:
			gtk.main()
		except KeyboardInterrupt:
			pass

	def quit(self, data=None):

		gtk.main_quit()


	def run_cmd(self, cmdlist):

		try:
			stdout = subprocess.check_output(cmdlist)
		except subprocess.CalledProcessError:
			   pass
		else:
			if stdout:
				return stdout


	def key_status(self):

		label = ''
		status = []
		keys = { '3':'C', '7':'N', '11':'S' }

		for line in self.run_cmd(['xset', 'q']).split('\n') :
			if 'Caps Lock:' in line:
				status = line.split()

		for index in 3, 7, 11:
			if status[index] == "on" :
			   label += " [" + keys[str(index)] + "] "

		return label


	def update_label(self):

		cwd = os.getcwd()
		red_icon = os.path.join(cwd, 'red.png')
		green_icon = os.path.join(cwd, 'green.png')
		label_text = self.key_status()
		if '[' in label_text:
			self.app.set_icon(red_icon)
		else:
			self.app.set_icon(green_icon)

		self.app.set_label(label_text)
		glib.timeout_add_seconds(1, self.set_app_label)


	def set_app_label(self):

		self.update_label()

def main():

	indicator = LockKeyStatusIndicator()
	indicator.run()

if __name__ == '__main__':
	main()


lks-indicator.py
#!/usr/bin/env python
# coding: utf-8

###########################################################
# Author: Serg Kolo , contact: 1047481448@qq.com
# Date: July 16, 2012
# Purpose: Simple indicator of Caps, Num, and Scroll Lock
#           keys for Ubuntu
#
# Written for: http://askubuntu.com/q/796985/295286
# Tested on: Ubuntu 16.04 LTS
#
###########################################################
#
# Licensed under The MIT License (MIT).
# See included LICENSE file or the notice below.
#
# Copyright © 2016 Sergiy Kolodyazhnyy
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from gi.repository import GLib as glib
from gi.repository import AppIndicator3  as appindicator
from gi.repository import Gtk as gtk
import os
import subprocess

class LockKeyStatusIndicator(object): 

    def __init__(self): 

        self.app = appindicator.Indicator.new('LKS', "",
                      appindicator.IndicatorCategory.APPLICATION_STATUS)

        self.app.set_status(appindicator.IndicatorStatus.ACTIVE)

        self.update_label()

        self.app_menu = gtk.Menu()
        self.quit_app = gtk.MenuItem('Quit')
        self.quit_app.connect('activate', self.quit)
        self.quit_app.show()
        self.app_menu.append(self.quit_app)

        self.app.set_menu(self.app_menu)

    def run(self): 

        try: 
            gtk.main()
        except KeyboardInterrupt: 
            pass

    def quit(self, data=None): 

        gtk.main_quit()


    def run_cmd(self, cmdlist): 

        try:
            stdout = subprocess.check_output(cmdlist)
        except subprocess.CalledProcessError:
               pass
        else:
            if stdout:
                return stdout


    def key_status(self):

        label = ''
        status = []
        keys = { '3':'C', '7':'N', '11':'S' }
        for line in self.run_cmd(['xset', 'q']).decode().split('\n') :
            if 'Caps Lock:' in line:
                status = line.split()

        for index in 3, 7, 11:
                        # TODO: add option to ignore keys , pass as 
                        # parameter to this function
                        #if keys[ str(index)  ] == ignore_keys:
                        #   continue
                        if status[index] == "on":
                           label += " [" + keys[str(index)] + "] "

        return label


    def update_label(self):

        cwd = os.getcwd()
        red_icon = os.path.join(cwd, 'red.png')
        green_icon = os.path.join(cwd, 'green.png')
        label_text = self.key_status()
        if '[' in label_text:
            self.app.set_icon(red_icon)
        else:
            self.app.set_icon(green_icon)

        self.app.set_label(label_text,"")
        glib.timeout_add_seconds(1, self.set_app_label)


    def set_app_label(self):

        self.update_label()

def main():

    indicator = LockKeyStatusIndicator()
    indicator.run()

if __name__ == '__main__':
    main()
        
